contracts/exchangeIssuance/ExchangeIssuanceLeveraged.sol

```diff
@@ -0,0 +1,926 @@
+        _initiateRedemption(_setToken, _amountSetToken, _exchange, PaymentToken.ERC20, paymentParams);
+    }
+
+    /**
+     * Trigger issuance of set token paying with any arbitrary ERC20 token
+     *
+     * @param _setToken            Set token to issue
+     * @param _amountSetToken      Amount to issue
+     * @param _maxAmountInputToken Maximum amount of input token to spend
+     * @param _exchange            Exchange to use in swap from short to long token
+     */
+    function issueExactSetFromERC20(
+        ISetToken _setToken,
+        uint256 _amountSetToken,
+        address _inputToken,
+        uint256 _maxAmountInputToken,
+        Exchange _exchange
+    )
+        isSetToken(_setToken)
+        external
+        nonReentrant
+    {
+        bytes memory paymentParams = abi.encode(_maxAmountInputToken, _inputToken);
+        _initiateIssuance(_setToken, _amountSetToken, _exchange, PaymentToken.ERC20, paymentParams);
+    }
+
+    /**
+     * Trigger issuance of set token paying with Eth
+     *
+     * @param _setToken            Set token to issue
+     * @param _amountSetToken      Amount to issue
+     * @param _exchange            Exchange to use in swap from short to long token
+     */
+    function issueExactSetFromETH(
+        ISetToken _setToken,
+        uint256 _amountSetToken,
+        Exchange _exchange
+    )
+        isSetToken(_setToken)
+        external
+        payable
+        nonReentrant
+    {
+        bytes memory paymentParams = abi.encode(msg.value);
+        _initiateIssuance(_setToken, _amountSetToken, _exchange, PaymentToken.ETH, paymentParams);
+    }
+
+    /**
+     * This is the callback function that will be called by the AaveLending Pool after flashloaned tokens have been sent
+     * to this contract.
+     * After exiting this function the Lending Pool will attempt to transfer back the loaned tokens + interest. If it fails to do so
+     * the whole transaction gets reverted
+     *
+     * @param assets     Addresses of all assets that were borrowed
+     * @param amounts    Amounts that were borrowed
+     * @param premiums   Interest to be paid on top of borrowed amount
+     * @param params     Encoded bytestring of other parameters from the original contract call to be used downstream
+     */
+    function executeOperation(
+        address[] calldata assets,
+        uint256[] calldata amounts,
+        uint256[] calldata premiums,
+        address , 
+        bytes calldata params
+    ) external override  onlyLendingPool returns (bool) {
+        require(assets.length == 1, "Exchange Issuance Leveraged: TOO MANY ASSETS");
+        require(amounts.length == 1, "Exchange Issuance Leveraged: TOO MANY AMOUNTS");
+        require(premiums.length == 1, "Exchange Issuance Leveraged: TOO MANY PREMIUMS");
+
+        DecodedParams memory decodedParams = abi.decode(params, (DecodedParams));
+        if(decodedParams.isIssuance){
```

Comment:
Can we add some comments on why we are doing each step for both the issue and redeem? Its hard to understand exactly whats going here without closely reading what each function does. Maybe the naming of these functions can be improved too?

Reply:
Added comments and removed the `_obtainShortTokens`  method since it became a pointless wrapper around `_swapLongForShortToken`. Imo it should be understandable with the comments but I'll review function names later and see if I can come up with better ones.

Reply:
I ended up refactoring this part rather extensively, hopefully increasing readability. If you still think this is  lacking in that regard please let me know.